symDiskCache.py

```diff
@@ -0,0 +1,97 @@
+from symUtil import mkdir_p
+import cPickle as pickle
+import os
+from symLogging import LogDebug
+from collections import deque
+
+# Manage in disk symbol cache
+class SymDiskCache(object):
+  def __init__(self, cachePath, maxNumberOfFiles):
+    mkdir_p(cachePath)
+    self.cachePath = cachePath
+    self.maxNumberOfFiles = maxNumberOfFiles
+
+    self.entryCount, fileList = self._Walk(cachePath)
+
+    # sort the symbol relative to the last access time
+    fileList.sort(cmp=lambda a,b: cmp(a[1], b[1]))
+
+    # filter out access times
+    self.lru = deque(map(lambda x: x[0], fileList))
+
+    self.CheckCacheSize()
+
+  def Fetch(self, filename):
+    r"fetch a symbol file from the cache"
+    path = self.absPath(filename)
+    symbolInfo = None
+    if os.path.exists(path):
+      with open(path, 'r') as f:
+        symbolInfo = pickle.load(f)
+        self.Use(filename)
+    return symbolInfo
+
+  def Store(self, symbolInfo, filename):
+    r"""Store a symbol file in the cache.
+
+    The caller must call use(filename) before this
+    method.
+    """
+    path = self.absPath(filename)
+    dirname = os.path.dirname(path)
+    mkdir_p(dirname)
+    self.Add(filename)
+
+    with open(path, 'w') as f:
+      pickle.dump(symbolInfo, f, pickle.HIGHEST_PROTOCOL)
+
+    self.entryCount += 1
+    self.CheckCacheSize()
```

Comment:
Think about what kind of eviction pattern this is going to create. Let's say the limit is 50 files. The server is going to cache 50 files and then get a request that caches 5 more files and will have to delete  the 5 oldest files. Then it will get another request and cache 7 new files and delete the 7 oldest again. And so on.

So after the cache reaches quota, every single request will have to pay the cost of evicting a whole bunch of files. That's not ideal.

How about when a request comes in that would cause to go over quota, you enqueue a new event to do quota maintenance. The maintenance event should delete say 40% of all files (based on age). You should also have a timer event to do the quota maintenance periodically (every 8 hours?).